kubo-rpc/src/http/swarm.rs

```diff
@@ -64,35 +63,92 @@ where
+impl<'de> Deserialize<'de> for ConnectQuery {
+    fn deserialize<D>(deserializer: D) -> Result<ConnectQuery, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct FieldVisitor;
+
+        impl<'de> serde::de::Visitor<'de> for FieldVisitor {
+            type Value = ConnectQuery;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str("`arg`")
+            }
+
+            fn visit_map<V>(self, mut map: V) -> Result<ConnectQuery, V::Error>
+            where
+                V: serde::de::MapAccess<'de>,
+            {
+                let mut args: Vec<String> = Vec::default();
+                while let Some(key) = map.next_key()? {
+                    match key {
+                        "arg" => args.push(map.next_value::<String>()?),
+                        _ => unreachable!(),
+                    }
+                }
+                Ok(ConnectQuery { arg: args })
+            }
+        }
+        deserializer.deserialize_identifier(FieldVisitor)
+    }
 }
 
-#[tracing::instrument(skip(data))]
 async fn connect<T>(
     data: web::Data<AppState<T>>,
     query: web::Query<ConnectQuery>,
 ) -> Result<HttpResponse, Error>
 where
     T: IpfsDep,
 {
-    let ma = Multiaddr::from_str(query.arg.as_str()).map_err(|e| Error::Invalid(e.into()))?;
-    let mh = ma
+    // Iterate over each arg and parse it as a multiaddr and search for peer ids.
+    let iter = query
+        .arg
         .iter()
-        .flat_map(|proto| {
-            if let Protocol::P2p(mh) = proto {
-                vec![mh]
-            } else {
-                vec![]
-            }
-        })
-        .next()
-        .ok_or_else(|| Error::Invalid(anyhow!("multiaddr does not contain p2p peer Id")))?;
-    let peer_id =
-        PeerId::from_multihash(mh).map_err(|_e| Error::Invalid(anyhow!("invalid peer Id")))?;
+        .map(|addr| -> Result<(Multiaddr, Option<PeerId>), Error> {
+            let addr = Multiaddr::from_str(addr).map_err(|e| Error::Invalid(e.into()))?;
+            let peer_id = addr
+                .iter()
+                .flat_map(|proto| match proto {
+                    Protocol::P2p(mh) => vec![mh],
+                    _ => Vec::new(),
+                })
+                .next()
+                .map(|mh| -> Result<PeerId, Error> {
+                    PeerId::from_multihash(mh)
+                        .map_err(|_e| Error::Invalid(anyhow!("invalid peer Id")))
+                })
+                .transpose()?;
+            Ok((addr, peer_id))
+        });
+
+    let (addrs, peer_ids) =
+        itertools::process_results(iter, |iter| iter.unzip::<_, _, Vec<_>, Vec<_>>())?;
+
+    // Check we found exactly one unique peer id.
+    let peer_ids: HashSet<PeerId> = peer_ids
+        .into_iter()
+        .flat_map(|p| if let Some(p) = p { vec![p] } else { Vec::new() })
+        .collect();
+    let peer_id = match peer_ids.len() {
```

Comment:
`split_first` instead?

Reply:
What is `split_first`?

Reply:
I prefer the current implementation, as peer_ids is not a slice and it requires a few step to create a slice from the set that the current logic is the simpler option. And the match on the length makes it very clear which each case means.